geomesa-spark/geomesa-spark-sql/src/test/scala/org/locationtech/geomesa/spark/SparkUtilsTest.scala

```diff
@@ -0,0 +1,211 @@
+
+  import scala.collection.JavaConverters._
+
+  sequential
+
+  var spark: SparkSession = _
+  var sc: SQLContext = _
+
+  val simpleTypeValueMap: Map[String, AnyRef] = Map(
+    "Integer" -> Int.box(10),
+    "Long" -> Long.box(42),
+    "Float" -> Float.box(3.14f),
+    "Double" -> Double.box(42.10),
+    "String" -> "test_string",
+    "Boolean" -> Boolean.box(true),
+    "Date" -> new java.util.Date(System.currentTimeMillis()),
+    "Bytes" -> Array.fill(10)((Random.nextInt(256) - 128).toByte)
+  )
+
+  val geomTypeValueMap: Map[String, AnyRef] = Map(
+    "Point" -> FastConverter.convert(
+      "POINT(45.0 49.0)",
+      classOf[org.locationtech.jts.geom.Point]),
+    "LineString" -> FastConverter.convert(
+      "LINESTRING(0 2, 2 0, 8 6)", classOf[org.locationtech.jts.geom.LineString]),
+    "Polygon" -> FastConverter.convert(
+      "POLYGON((20 10, 30 0, 40 10, 30 20, 20 10))", classOf[org.locationtech.jts.geom.Polygon]),
+    "MultiPoint" -> FastConverter.convert(
+      "MULTIPOINT(0 0, 2 2)", classOf[org.locationtech.jts.geom.MultiPoint]),
+    "MultiLineString" -> FastConverter.convert(
+      "MULTILINESTRING((0 2, 2 0, 8 6),(0 2, 2 0, 8 6))", classOf[org.locationtech.jts.geom.MultiLineString]),
+    "MultiPolygon" -> FastConverter.convert(
+      "MULTIPOLYGON(((-1 0, 0 1, 1 0, 0 -1, -1 0)), ((-2 6, 1 6, 1 3, -2 3, -2 6)), ((-1 5, 2 5, 2 2, -1 2, -1 5)))",
+      classOf[org.locationtech.jts.geom.MultiPolygon]),
+    "GeometryCollection" -> FastConverter.convert(
+      "GEOMETRYCOLLECTION(POINT(45.0 49.0),POINT(45.1 49.1))",
+      classOf[org.locationtech.jts.geom.GeometryCollection]),
+    "Geometry" -> FastConverter.convert(
+      "POINT(45.0 49.0)",
+      classOf[org.locationtech.jts.geom.Point])
+  )
+
+  val unsupportedTypeValueMap: Map[String, AnyRef] = Map(
+    "UUID" -> UUID.randomUUID()
+  )
+
+  def validateDataTypeConversions(sf: SimpleFeature): Unit = {
+    val sft = sf.getFeatureType
+    val featureNames = sft.getAttributeDescriptors.asScala.map(d => d.getLocalName)
+    val schema = SparkUtils.createStructType(sft)
+    val extractors = SparkUtils.getExtractors(schema.fieldNames, schema)
+    val row = SparkUtils.sf2row(schema, sf, extractors)
+    validateRow(sf, row)
+
+    // create dataframe to validate correctness of schema
+    val df = spark.createDataFrame(spark.sparkContext.parallelize(Seq(row)), schema)
+    val row2 = df.collect()(0)
+    validateRow(sf, row2)
+
+    // simple feature should preserve its value when converted back from Row
+    val sf2 = SparkUtils.rowsToFeatures(sft, schema).apply(row2)
+    featureNames.foreach { f =>
+      val attrType = sft.getDescriptor(f).getType
+      if (attrType.getBinding == classOf[java.util.List[_]]) {
+        sf.getAttribute(f).asInstanceOf[java.util.List[_]].toArray mustEqual
+          sf2.getAttribute(f).asInstanceOf[java.util.List[_]].toArray
+      } else if (attrType.getBinding == classOf[java.util.Map[_, _]]) {
+        ok
```

Comment:
could we add a comparison here? probably just a scala.collection.JavaConverters `.asScala` would allow it to work

Reply:
Sure, I've added the check here. Since we have byte arrays inside maps, we have to try harder to make specs2 looking into the values of the maps.